File: demo/scripts/example-clean-script.py
#!/usr/bin/env python3
"""Example: a clean Python script that passes all quality gates.

Quality gates this script demonstrably satisfies:
  - Cyclomatic complexity (radon CC) < 6 for every function
  - Full type annotations (PEP 604 + built-in generics, mypy --strict clean)
  - No bare/broad except clauses (ruff E722 + F841 compliant)
  - No mutable default arguments
  - All file I/O uses context managers (no unsafe open())
  - No swallowed errors — exceptions are re-raised or logged with context
"""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Sequence


def read_file_lines(path: Path) -> list[str]:
    """Return all non-empty lines from *path*, stripped of trailing whitespace."""
    with path.open("r", encoding="utf-8") as handle:
        return [line.rstrip() for line in handle if line.strip()]


def filter_comment_lines(lines: list[str]) -> list[str]:
    """Filter out lines that start with ``#``."""
    return [line for line in lines if not line.lstrip().startswith("#")]


def count_words(lines: Sequence[str]) -> dict[str, int]:
    """Return a word-frequency dictionary from the given *lines*."""
    counts: dict[str, int] = {}
    for line in lines:
        for word in line.split():
            counts[word] = counts.get(word, 0) + 1
    return counts


def top_n_words(counts: dict[str, int], n: int) -> list[tuple[str, int]]:
    """Return the *n* most common ``(word, frequency)`` pairs."""
    sorted_items = sorted(counts.items(), key=lambda pair: pair[1], reverse=True)
    return sorted_items[:n]


def analyze_text_file(path: str, top_n: int = 10) -> None:
    """Read, analyze, and print a word-frequency summary for *path*."""
    file_path = Path(path)
    if not file_path.is_file():
        print(f"Error: {file_path} is not a regular file.", file=sys.stderr)
        sys.exit(1)

    lines = read_file_lines(file_path)
    code_lines = filter_comment_lines(lines)
    counts = count_words(code_lines)
    top = top_n_words(counts, top_n)

    print(f"Analyzed {len(code_lines)} non-comment lines in {file_path.name}")
    print("Top words:")
    for word, freq in top:
        print(f"  {word}: {freq}")


def main() -> int:
    """Entry point: parse arguments and run the analysis."""
    args = sys.argv[1:]
    if len(args) != 1:
        print(f"Usage: {Path(sys.argv[0]).name} <text-file-path>", file=sys.stderr)
        return 2

    try:
        analyze_text_file(args[0])
    except OSError as exc:
        print(f"File error: {exc}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: demo/scripts/example-clean-script.ts
#!/usr/bin/env node
/**
 * Example: a clean TypeScript script that passes all quality gates.
 *
 * Quality gates this script demonstrably satisfies:
 *   - Cyclomatic complexity < 6 for every function (ts/complexity [[2,5]])
 *   - Strict TSConfig types (tsc --strict, noImplicitAny, strictNullChecks)
 *   - No empty catch or swallowed errors (no-empty, no-useless-catch)
 *   - No mutable default arguments (TS avoids this via object.freeze or
 *     undefined defaults)
 *   - All file I/O uses synchronous calls with proper error context
 */

import { readFileSync } from "node:fs";
import { basename } from "node:path";
import { exit } from "node:process";

interface WordFrequency {
  word: string;
  count: number;
}

function readLines(filePath: string): string[] {
  const content: string = readFileSync(filePath, "utf-8");
  return content
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
}

function filterComments(lines: string[]): string[] {
  return lines.filter((line) => !line.startsWith("#"));
}

function countWords(lines: string[]): Map<string, number> {
  const counts = new Map<string, number>();
  for (const line of lines) {
    const words = line.split(/\s+/);
    for (const word of words) {
      if (word.length === 0) continue;
      const current = counts.get(word) ?? 0;
      counts.set(word, current + 1);
    }
  }
  return counts;
}

function topNWords(counts: Map<string, number>, n: number): WordFrequency[] {
  const entries = Array.from(counts.entries());
  entries.sort((a, b) => b[1] - a[1]);
  return entries.slice(0, n).map(([word, count]) => ({ word, count }));
}

function analyzeFile(filePath: string, topN: number = 10): void {
  const lines = readLines(filePath);
  const codeLines = filterComments(lines);
  const counts = countWords(codeLines);
  const top = topNWords(counts, topN);

  console.log(
    `Analyzed ${codeLines.length} non-comment lines in ${basename(filePath)}`,
  );
  console.log("Top words:");
  for (const { word, count } of top) {
    console.log(`  ${word}: ${count}`);
  }
}

function main(): number {
  const args = process.argv.slice(2);
  if (args.length !== 1) {
    console.error(`Usage: ${basename(process.argv[1])} <text-file-path>`);
    return 2;
  }

  const filePath = args[0] as string;
  try {
    analyzeFile(filePath);
  } catch (err: unknown) {
    if (err instanceof Error) {
      console.error(`File error: ${err.message}`);
    } else {
      console.error("Unknown error occurred");
    }
    return 1;
  }
  return 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  exit(main());
}

export { analyzeFile, countWords, filterComments, readLines, topNWords };
